Add unit tests for multiple python versions to CI stage

fpgmaas opened the pull request:
This PR builds upon [this PR](https://github.com/MrPowers/chispa/pull/80). When that is merged I can update this branch to better reflect the changes in "Files changed".

Related issue: https://github.com/MrPowers/chispa/issues/82

---

This PR adds testing against multiple Python versions, updates some of the mkdocs dependencies, and changes the trigger of the CI checks to run on each pull request and on each commit to master (i.e. merging a PR). For an example run, see [here](https://github.com/fpgmaas/chispa/actions/runs/6564589223), or approve the workflow run on this PR.

This also adds a check for the `poetry.lock` file; to prevent issues like #107 

I think this can be helpful, let me know what you think of the proposal!

MrPowers commented:
@fpgmaas - really sorry for never responding here.  Should we have merged this one?

fpgmaas commented:
@MrPowers  No worries; I think that is my own doing, since you pointed out in another PR that you did not want to drop support for Python 3.7 just yet, and then life got in the way and I didn't get around to updating it. I'll take a look at it again somewhere this week and try to modify & reopen PR's that could still be useful.

MrPowers commented:
@fpgmaas - yea, thanks for understanding.

We're actually dropping support for PySpark 2 on quinn, so now is probably a good time to make these changes for chispa as well.  Let's brainstorm the next steps for this project!

fpgmaas commented:
@MrPowers I updated [the PR](https://github.com/MrPowers/chispa/pull/80) that this one is built on top of and rebased this one. Example run can be found [here](https://github.com/fpgmaas/chispa/actions/runs/9930074399).

fpgmaas commented:
@SemyonSinchenko just a heads up that this is an old PR so it won't show up at the top of open PR's. If you have time, could you review? No hurry! :)

SemyonSinchenko reviewed (commented):
I left few comments

fpgmaas commented on .github/workflows/ci.yml:
```diff
@@ -1,39 +1,56 @@
 name: Unit tests
 
 on:
+  pull_request:
```
@SemyonSinchenko I believe that limit only holds for private repositories; public repositories have unlimited build minutes.

fpgmaas commented on .github/workflows/ci.yml:
```diff
@@ -1,39 +1,56 @@
-
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
```
We should definitely test for different Python versions to ensure that `chispa` works for each Python version. In addition we can also test for multiple versions of Pyspark, as I see you did in `quinn`. 

Ideally, we test for multipel versions of Python and Pyspark, similar to [here](https://github.com/fpgmaas/deptry/actions/runs/9938821113) with different OS and Python versions. Is it ok if we handle that in a separate PR? I will create an issue for this.

fpgmaas replied:
Created #109  and #110

fpgmaas commented on pyproject.toml:
```diff
@@ -8,22 +8,22 @@ license = "MIT"
 mkdocs-section-index = "^0.3.5"
 markdown-include = "^0.8.1"
-mkdocs = "^1.4.2"
+mkdocs = "^1.5.3"
```
Sounds good to me, I will change it accordingly.

fpgmaas commented:
> I left few comments

I think I resolved them!

SemyonSinchenko reviewed (approved):
LGTM, thank you!
